pkg/policy/rule_test.go

```diff
@@ -233,3 +233,83 @@ func (ds *PolicyTestSuite) TestL3Policy(c *C) {
 	}.Validate()
 	c.Assert(err, Not(IsNil))
 }
+
+func (ds *PolicyTestSuite) TestRuleCanReachFromEntity(c *C) {
+	fromWorld := &SearchContext{
+		From: labels.ParseSelectLabelArray("reserved:world"),
+		To:   labels.ParseSelectLabelArray("bar"),
+	}
+
+	notFromWorld := &SearchContext{
+		From: labels.ParseSelectLabelArray("foo"),
+		To:   labels.ParseSelectLabelArray("bar"),
+	}
+
+	rule1 := rule{
+		api.Rule{
+			EndpointSelector: api.NewESFromLabels(labels.ParseSelectLabel("bar")),
+			Ingress: []api.IngressRule{
+				{
+					FromEntities: []api.Entity{api.EntityWorld},
+				},
+			},
+		},
+	}
+
+	state := traceState{}
+	c.Assert(rule1.canReach(fromWorld, &state), Equals, api.Allowed)
+	c.Assert(state.selectedRules, Equals, 1)
+	state = traceState{}
+	c.Assert(rule1.canReach(notFromWorld, &traceState{}), Equals, api.Undecided)
+	c.Assert(state.selectedRules, Equals, 0)
+}
+
+func (ds *PolicyTestSuite) TestRuleCanReachEntity(c *C) {
+	toWorld := &SearchContext{
+		From: labels.ParseSelectLabelArray("bar"),
+		To:   labels.ParseSelectLabelArray("reserved:world"),
+	}
+
+	notToWorld := &SearchContext{
+		From: labels.ParseSelectLabelArray("bar"),
+		To:   labels.ParseSelectLabelArray("foo"),
+	}
+
+	rule1 := rule{
+		api.Rule{
+			EndpointSelector: api.NewESFromLabels(labels.ParseSelectLabel("bar")),
+			Egress: []api.EgressRule{
+				{
+					ToEntities: []api.Entity{api.EntityWorld},
```

Comment:
Can you extend this to also include an ingress `FromEndpoints` section so we can verify that ToEntities does not break the other rules?

Reply:
will do